083-stretch-cluster.md

```diff
@@ -0,0 +1,155 @@
+
+### Prerequisites
+
+- **Multiple Kubernetes Clusters**: Stretch Kafka clusters will require multiple Kubernetes clusters. Ideally, an odd number of clusters (at least three) is needed to maintain quorum in the event of a cluster outage.
+
+- **Low Latency**: Kafka clusters should be deployed in environments that allow low-latency communication between Kafka brokers and controllers. Stretch Kafka clusters should be deployed in environments such as data centers or availability zones within a single region, and not across distant regions where high latency could impair performance.
+
+- **KRaft**: As Kafka and Strimzi transition towards KRaft-based clusters, this proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters. While Zookeeper-based deployments are still supported, they are outside the scope of this proposal.
+
+### Design
+
+The cluster operator will be deployed in all Kubernetes clusters and will manage Kafka brokers/controllers running on that cluster. One Kubernetes cluster will act as the control point for defining custom resources (Kafka, KafkaNodePool) required for stretch Kafka cluster. The KafkaNodePool custom resource will be extended to include information about a Kubernetes cluster where the pool should be deployed. The cluster operator will create necessary resources (StrimziPodSets, services etc.) on the target clusters specified within the KafkaNodePool resource.
+
+This approach will allow users to specify/manage the definition of stretch Kafka cluster in a single location. The operators will then create necessary resources in target Kubernetes clusters, which can then be reconciled/managed by operators on those clusters.
+
+### Reconciling Kafka and KafkaNodePool resources
+<img width="1321" alt="image" src="https://media.github.ibm.com/user/473362/files/0a1a4e7f-8227-47da-86b1-4b27a83fdb9d">
+
+### Reconciling StrimziPodSet resources
+<img width="1317" alt="image" src="https://media.github.ibm.com/user/473362/files/f4524847-3c9c-4a8c-a5fd-4a76b6ec0bf4">
+
+#### KafkaNodePool changes
+A new optional field (`target`) will be introduced in the KafkaNodePool resource specification, to allow users to specify the details of the Kubernetes cluster where the node pool should be deployed. This section will include the target cluster's URL (Kubernetes cluster where resources for this node pool will be created) and the secret containing the kubeconfig data for that cluster. 
+
+An example of the KafkaNodePool resource with the new fields might look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  target:
+    clusterUrl: <K8S Cluster URL>
+    secret: <SecretName>
```

Comment:
Here’s what I understand from your suggestion. Is my understanding correct? 

### using env var (Option 1 )
Add the cluster information as env var in the CO Deployment

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: strimzi-cluster-operator
  namespace: myproject
spec:
  template:
    spec:
      containers:
        - name: strimzi-cluster-operator
          env:
            - name: CLUSTER_A_URL
              value: "<K8S_CLUSTER_A_URL>"
            - name: CLUSTER_A_SECRET_NAME
              value: "<SecretNameA>"
            - name: CLUSTER_B_URL
              value: "<K8S_CLUSTER_B_URL>"
            - name: CLUSTER_B_SECRET_NAME
              value: "<SecretNameB>"
         ................................
         ...............................

```

### (Option 2) Alternatively, we can define a CM for cluster URLs and a Secret for sensitive information like

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-info
  namespace: myproject
data:
  clusters.yaml: |
    clusters:
      - name: cluster-a
        url: <K8S_CLUSTER_A_URL>
      - name: cluster-b
        url: <K8S_CLUSTER_B_URL>
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cluster-secrets
  namespace: myproject
type: Opaque
data:
  cluster-a-secret: <base64-encoded-secret-for-cluster-a>
  cluster-b-secret: <base64-encoded-secret-for-cluster-b>
```
Then mount CM and Secret to CO deployment 

```yaml
spec:
  template:
    spec:
      containers:
        - name: strimzi-cluster-operator
          envFrom:
            - configMapRef:
                name: cluster-info
            - secretRef:
                name: cluster-secrets
```

Now reference clusters by alias in KNP CR like

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: controller
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  target:
    clusterAlias: "cluster-a"  # Referencing the alias defined in CO
```

Reply:
Well, when using the environment variable, you should IMHO consider using a single environment variable with some map similar to what we use for images.

But the main problem is that we cannot decide on some _API change_ before knowing the implementation details. The best way to configure it depends on how it will work:
* How will users create these accounts on various Kube distributions? How will it work on OpenShift? how on AKS EKS? GKE? Rancher?
* What are these _credentials_? Are they `kubeconfig` files? Should they be an API Server URL + token?
* Are these long-term credentials? Short-term credentials that will be changing?
* How will these credentials be used in the code to create and share the clients for the different clusters?
* How will the RBACs of these clients look like on the remote clusters?

That needs to be clarified and designed. And that should drive the optimal outcome of how the API will look like.